### PR TITLE
[Merged by Bors] - feat: port Algebra.GradedMonoid

### PR DESCRIPTION
**Pull request by mattrobball**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mattrobball** [comment]
This is my first effort at porting so be warned.

**Comment by dupuisf** [comment]
The PR currently makes all kinds of changes to `lake-manifest.json` (probably from running `lake update`); try to revert those and make sure not to touch this file.

File: Mathlib/Algebra/GradedMonoid.lean [hunk 1]
```diff
@@ -0,0 +1,717 @@
...
+    (GradedMonoid.mk _ <| gnpowRec n.succ a.snd) = a * ⟨_, gnpowRec n a.snd⟩ :=
+  Sigma.ext (succ_nsmul _ _) (heq_of_cast_eq _ rfl).symm
+#align graded_monoid.gmonoid.gnpow_rec_succ GradedMonoid.GMonoid.gnpow_rec_succ
+
+end GMonoid
+
+/-- A tactic to for use as an optional value for `Gmonoid.gnpow_zero'' -/
+macro "apply_gmonoid_gnpow_rec_zero_tac" : tactic => `(tactic | apply GMonoid.gnpow_rec_zero)
+/-- A tactic to for use as an optional value for `Gmonoid.gnpow_succ'' -/
+macro "apply_gmonoid_gnpow_rec_succ_tac" : tactic => `(tactic | apply GMonoid.gnpow_rec_succ)
```
Comment (dupuisf) [inline Mathlib/Algebra/GradedMonoid.lean:176]:
```suggestion
/-- A tactic to for use as an optional value for `GMonoid.gnpow_zero'' -/
macro "apply_gmonoid_gnpowRec_zero_tac" : tactic => `(tactic | apply GMonoid.gnpowRec_zero)
/-- A tactic to for use as an optional value for `GMonoid.gnpow_succ'' -/
macro "apply_gmonoid_gnpowRec_succ_tac" : tactic => `(tactic | apply GMonoid.gnpowRec_succ)
```

File: Mathlib/Algebra/GradedMonoid.lean [hunk 2]
```diff
@@ -0,0 +1,717 @@
...
+
+section DProd
+
+variable {α : Type _} {A : ι → Type _} [AddMonoid ι] [GradedMonoid.GMonoid A]
+
+/-- The index used by `List.dprod`. Propositionally this is equal to `(l.map fι).Sum`, but
+definitionally it needs to have a different form to avoid introducing `Eq.rec`s in `List.dprod`. -/
+def List.dProdIndex (l : List α) (fι : α → ι) : ι :=
+  l.foldr (fun i b => fι i + b) 0
+#align list.dprod_index List.dProdIndex
```
Comment (dupuisf) [inline Mathlib/Algebra/GradedMonoid.lean:364]:
```suggestion
def List.dprodIndex (l : List α) (fι : α → ι) : ι :=
  l.foldr (fun i b => fι i + b) 0
#align list.dprod_index List.dProdIndex
```

File: Mathlib/Algebra/GradedMonoid.lean [hunk 3]
```diff
@@ -0,0 +1,717 @@
...
+
+@[simp]
+theorem List.dprod_index_nil (fι : α → ι) : ([] : List α).dProdIndex fι = 0 :=
```
Comment (dupuisf) [inline Mathlib/Algebra/GradedMonoid.lean:367]:
```suggestion
theorem List.dprodIndex_nil (fι : α → ι) : ([] : List α).dProdIndex fι = 0 :=
```
and the same throughout

File: Mathlib/Algebra/GradedMonoid.lean [hunk 4]
```diff
@@ -0,0 +1,717 @@
...
+instance CommMonoid.gCommMonoid [AddCommMonoid ι] [CommMonoid R] :
+    GradedMonoid.GCommMonoid fun _ : ι => R :=
+  { Monoid.gMonoid ι with
+    mul_comm := fun _ _ => Sigma.ext (add_comm _ _) (heq_of_eq (mul_comm _ _)) }
+#align comm_monoid.gcomm_monoid CommMonoid.gCommMonoid
+
+/-- When all the indexed types are the same, the dependent product is just the regular product. -/
+@[simp]
+theorem List.dprod_monoid {α} [AddMonoid ι] [Monoid R] (l : List α) (fι : α → ι) (fA : α → R) :
+    -- (l.dprod fι fA : (fun i : ι => R) _ ) = ((l.map fA).prod : R) :=
```
Comment (dupuisf) [inline Mathlib/Algebra/GradedMonoid.lean:473]: I guess this should be deleted?

File: Mathlib/Algebra/GradedMonoid.lean [hunk 5]
```diff
@@ -0,0 +1,717 @@
...
+#align set_like.one_mem_graded SetLike.one_mem_graded
+
+instance SetLike.gOne {S : Type _} [SetLike S R] [One R] [Zero ι] (A : ι → S)
+    [SetLike.GradedOne A] : GradedMonoid.GOne fun i => A i
+    where one := ⟨1, SetLike.one_mem_graded _⟩
+#align set_like.ghas_one SetLike.gOne
+
+@[simp]
+theorem SetLike.coe_ghas_one {S : Type _} [SetLike S R] [One R] [Zero ι] (A : ι → S)
+    [SetLike.GradedOne A] : ↑(@GradedMonoid.GOne.one _ (fun i => A i) _ _) = (1 : R) :=
```
Comment (dupuisf) [inline Mathlib/Algebra/GradedMonoid.lean:513]:
```suggestion
theorem SetLike.coe_gOne {S : Type _} [SetLike S R] [One R] [Zero ι] (A : ι → S)
    [SetLike.GradedOne A] : ↑(@GradedMonoid.GOne.one _ (fun i => A i) _ _) = (1 : R) :=
```

File: Mathlib/Algebra/GradedMonoid.lean [hunk 6]
```diff
@@ -0,0 +1,717 @@
...
+  SetLike.GradedMul.mul_mem hi hj
+#align set_like.mul_mem_graded SetLike.mul_mem_graded
+
+instance SetLike.gMul {S : Type _} [SetLike S R] [Mul R] [Add ι] (A : ι → S)
+    [SetLike.GradedMul A] : GradedMonoid.GMul fun i => A i
+    where mul := fun a b => ⟨(a * b : R), SetLike.mul_mem_graded a.prop b.prop⟩
+#align set_like.ghas_mul SetLike.gMul
+
+@[simp,nolint simpNF] 
+theorem SetLike.coe_ghas_mul {S : Type _} [SetLike S R] [Mul R] [Add ι] (A : ι → S)
```
Comment (dupuisf) [inline Mathlib/Algebra/GradedMonoid.lean:534]:
```suggestion
theorem SetLike.coe_gMul {S : Type _} [SetLike S R] [Mul R] [Add ι] (A : ι → S)
```

File: Mathlib/Algebra/GradedMonoid.lean [hunk 7]
```diff
@@ -0,0 +1,717 @@
...
+
+variable {R S : Type _} [SetLike S R]
+
+/-- An element `a : R` is said to be homogeneous if there is some `i : ι` such that `a ∈ A i`. -/
+def SetLike.IsHomogeneous (A : ι → S) (a : R) : Prop :=
+  ∃ i, a ∈ A i
+#align set_like.is_homogeneous SetLike.IsHomogeneous
+
+@[simp]
+theorem SetLike.is_homogeneous_coe {A : ι → S} {i} (x : A i) : SetLike.IsHomogeneous A (x : R) :=
```
Comment (dupuisf) [inline Mathlib/Algebra/GradedMonoid.lean:694]:
```suggestion
theorem SetLike.isHomogeneous_coe {A : ι → S} {i} (x : A i) : SetLike.IsHomogeneous A (x : R) :=
```
and so on

**Review by dupuisf — COMMENTED** [review]
There are a few cosmetic issues, but it's in good shape!

**Comment by mattrobball** [comment]
Thanks for pointing out naming convention inconsistency. That should be cleaned up now. I restored the existing lake-manifest. Not sure why this isn't in .gitignore.

File: Mathlib/Algebra/GradedMonoid.lean [hunk 8]
```diff
@@ -0,0 +1,715 @@
...
+
+/-- A variant of `GradedMonoid.mk_list_dProd` for rewriting in the other direction. -/
+theorem GradedMonoid.list_prod_map_eq_dProd (l : List α) (f : α → GradedMonoid A) :
+    (l.map f).prod = GradedMonoid.mk _ (l.dProd (fun i => (f i).1) fun i => (f i).2) :=
+  by
+  rw [GradedMonoid.mk_list_dProd, GradedMonoid.mk]
+  simp_rw [Sigma.eta]
+#align graded_monoid.list_prod_map_eq_dprod GradedMonoid.list_prod_map_eq_dProd
+
+theorem GradedMonoid.list_prod_of_fn_eq_dProd {n : ℕ} (f : Fin n → GradedMonoid A) :
```
Comment (jcommelin) [inline Mathlib/Algebra/GradedMonoid.lean:424]:
```suggestion
theorem GradedMonoid.list_prod_ofFn_eq_dProd {n : ℕ} (f : Fin n → GradedMonoid A) :
```

File: Mathlib/Algebra/GradedMonoid.lean [hunk 9]
```diff
@@ -0,0 +1,715 @@
...
+    rw [List.map_nil, List.map_nil, List.prod_nil, List.sum_nil]
+    exact one_mem_graded _
+  | head::tail => 
+    rw [List.map_cons, List.map_cons, List.prod_cons, List.sum_cons]
+    exact
+      mul_mem_graded (h _ <| List.mem_cons_self _ _)
+        (list_prod_map_mem_graded tail _ _ <| fun j hj => h _ <| List.mem_cons_of_mem _ hj)
+#align set_like.list_prod_map_mem_graded SetLike.list_prod_map_mem_graded
+
+theorem list_prod_of_fn_mem_graded {n} (i : Fin n → ι) (r : Fin n → R) (h : ∀ j, r j ∈ A (i j)) :
```
Comment (jcommelin) [inline Mathlib/Algebra/GradedMonoid.lean:586]:
```suggestion
theorem list_prod_ofFn_mem_graded {n} (i : Fin n → ι) (r : Fin n → R) (h : ∀ j, r j ∈ A (i j)) :
```

**Comment by mattrobball** [comment]
Went through the theorems again to check names4 and hopefully this is it. Includes fixes suggested by @jcommelin and some missed (??) by @dupuisf. I appreciate the feedback.

File: Mathlib/Algebra/GradedMonoid.lean (hunk 2 above)
Comment (jcommelin) [reply Mathlib/Algebra/GradedMonoid.lean:364]: @mattrobball This one is still open. Please replace all the `dProd`s by `dprod`.

Comment (mattrobball) [reply Mathlib/Algebra/GradedMonoid.lean:364]: I thought dependent product shortens to DProd when a type and dProd when a term. Does the not agree with naming4?

Comment (mattrobball) [reply Mathlib/Algebra/GradedMonoid.lean:364]: I have some other naming in the file based off this understanding of the new rules.

File: Mathlib/Algebra/GradedMonoid.lean [hunk 10]
```diff
@@ -491,52 +490,66 @@ section Subobjects
...
-/-- A version of `graded_monoid.gmonoid` for internally graded objects. -/
+"Left-hand side does not simplify, when using the simp lemma on itself."
+
+However, simp does indeed solve the following. Possibly related std#71,std#78
+
+example {S : Type _} [SetLike S R] [Mul R] [Add ι] (A : ι → S)
+    [SetLike.GradedMul A] {i j : ι} (x : A i) (y : A j) :
+    ↑(@GradedMonoid.GMul.mul _ (fun i => A i) _ _ _ _ x y) = (x * y : R) := by simp 
+
+-/
```
Comment (jcommelin) [inline Mathlib/Algebra/GradedMonoid.lean:548]: Please move the porting note to just above where you added `nolint simpNF`. And please mention `simpNF` in the line that says `Porting note`. That way grepping for `Porting note` gives an indication what the note is about.

File: Mathlib/Algebra/GradedMonoid.lean [hunk 11]
```diff
@@ -547,94 +560,122 @@ variable {A : ι → S} [SetLike.GradedMonoid A]
...
+Porting note: Linter returns 
+
+"Left-hand side does not simplify, when using the simp lemma on itself."
+
+However, simp does indeed solve the following. Possibly related std#71,std#78
+
+example {S : Type _} [SetLike S R] [Monoid R] [AddMonoid ι] (A : ι → S)
+    [SetLike.GradedMonoid A] {i : ι} (x : A i) (n : ℕ) :
+    ↑(@GradedMonoid.GMonoid.gnpow _ (fun i => A i) _ _ n _ x) = (x:R)^n := by simp
+
```
Comment (jcommelin) [inline Mathlib/Algebra/GradedMonoid.lean:626]: Idem

File: Mathlib/Algebra/GradedMonoid.lean [hunk 12]
```diff
@@ -547,94 +560,122 @@ variable {A : ι → S} [SetLike.GradedMonoid A]
...
 
-include R
+"Left-hand side does not simplify, when using the simp lemma on itself."
+
+However, simp does indeed solve the following. Possibly related std#71,std#78
+
+example (A : ι → S) [SetLike.GradedMonoid A] (fι : α → ι)
+    (fA : ∀ a, A (fι a)) (l : List α) : ↑(@List.dProd _ _ (fun i => ↥(A i)) _ _ l fι fA) 
+    = (List.prod (l.map fun a => fA a) : R) := by simp
+-/
```
Comment (jcommelin) [inline Mathlib/Algebra/GradedMonoid.lean:665]: Idem

**Review by jcommelin — COMMENTED** [review]
Thanks! Please moving some porting notes a tiny bit up, and then it's good to go!

bors d+

File: Mathlib/Algebra/GradedMonoid.lean (hunk 2 above)
Comment (jcommelin) [reply Mathlib/Algebra/GradedMonoid.lean:364]: aah, my bad. You're right!

File: Mathlib/Algebra/GradedMonoid.lean (hunk 10 above)
Comment (mattrobball) [reply Mathlib/Algebra/GradedMonoid.lean]: Done!

**Comment by mattrobball** [comment]
Thanks @dupuisf and @jcommelin for the help!

**Comment by mattrobball** [comment]
bors merge

**Pull request closed without merging**